plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +601,158 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String bucketName = getBucketName(uri);
+        String key = keyFromPath(path);
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
+        }
+
+        if (recursive) {
+            boolean deleted = false;
+            if (s3ObjectMetadata == null ||
```

Comment:
The condition here seems wrong, what if this is called with `s3://bucket/directory/file.txt` ? I would expect `s3DirectoryMetadata` (resulting from a call to `getObjectMetadata("s3://bucket/directory/file.txt/")` which returns null) to be null and therefore fall into a `deletePrefix` instead of throwing an exception (it's not a "directory" in this scenario).